Make react-hs compile with GHC

So currently I can't use ghci or hdevtools with react-hs.

I can't use ghci since ghcjs doesn't support it, or at least the version of ghcjs version specified [here](https://docs.haskellstack.org/en/stable/ghcjs/) doesn't.

I can't use hdevtools since hdevtools doesn't support ghcjs.

I am trying to figure out how to get a ghcjs version with ghci that works with stack, and I also added an issue to hdevtools for ghcjs support. But I do think that react-hs should support ghc.

A library that might help a lot is [ghcjs-base-stub](https://www.stackage.org/lts-8.12/package/ghcjs-base-stub-0.1.0.2).

So with my change it looks like there have been no degradations in the sense that it still compiles with GHCJS just fine, however it still does not compile with GHC. It seems like it is a step in the right direction, but it still gives me the following error message, and I have not interacted enough with JS FFI / FFI in general to know how to fix it:

```
 Configuring react-hs-0.0.1...
    Preprocessing library react-hs-0.0.1...
    [ 1 of 12] Compiling React.Flux.Internal ( src/React/Flux/Internal.hs, .stack-work/dist/x86_64-osx/Cabal-1.24.2.0/build/React/Flux/Internal.o )
    
    .../.stack-work/downloaded/ufTvVgsUDzR4/react-hs/src/React/Flux/Internal.hs:612:1: error:
        • The `javascript' calling convention is unsupported on this platform
        • When checking declaration:
            foreign import javascript unsafe "static $2[$1]" js_findFromArray
              :: Int -> JSVal -> JSVal
    
    .../.stack-work/downloaded/ufTvVgsUDzR4/react-hs/src/React/Flux/Internal.hs:612:1: error:
        • ‘$2[$1]’ is not a valid C identifier
        • When checking declaration:
            foreign import javascript unsafe "static $2[$1]" js_findFromArray
              :: Int -> JSVal -> JSVal
```

Any idea how to fix this issue?

I have unfortunately not looked into glazier-react at all. Could potentially look into it later.

Also do you have a contact email? I don't see one attached to your GitHub profile.

--- a/react-hs/react-hs.cabal
+++ b/react-hs/react-hs.cabal
@@ -69,7 +69,6 @@ library
                       TypeOperators
 
   build-depends: base >=4.9 && < 5
-               , ghcjs-base
                , deepseq
                , mtl >= 2.1
                , aeson >= 0.8
@@ -79,6 +78,10 @@ library
                , bytestring
                , template-haskell >= 2.10
                , string-conversions >= 0.4 && < 0.5
+  if impl(ghcjs)
+    build-depends: ghcjs-base
+  if !impl(ghcjs)
+    build-depends: ghcjs-base-stub
 
 executable test-client
    if !flag(test-client)

--- a/react-hs/react-hs.cabal
+++ b/react-hs/react-hs.cabal
@@ -81,7 +81,7 @@ library
   if impl(ghcjs)
     build-depends: ghcjs-base
   if !impl(ghcjs)
-    build-depends: ghcjs-base-stub
+    build-depends: ghcjs-base-stub 
 
 executable test-client
    if !flag(test-client)
@@ -93,9 +93,12 @@ executable test-client
    hs-source-dirs: test/client
    main-is: TestClient.hs
    build-depends: base
-                , ghcjs-base
                 , react-hs
                 , time
                 , text
                 , deepseq
                 , aeson
+  if impl(ghcjs)
+    build-depends: ghcjs-base
+  if !impl(ghcjs)
+    build-depends: ghcjs-base-stub 

--- a/react-hs/src/React/Flux.hs
+++ b/react-hs/src/React/Flux.hs
@@ -58,7 +58,9 @@
 -- The file <https://bitbucket.org/wuzzeb/react-flux/src/tip/test/spec/TodoSpec.hs test\/spec\/TodoSpec.hs>
 -- in the source code contains a hspec-webdriver test for the TODO example application.
 
+{-# LANGUAGE CPP #-}
 {-# OPTIONS_GHC -fno-warn-duplicate-exports #-} -- ArgumentsToProps is exported twice, once by React.Flux.PropertiesAndEvents and once here
+
 module React.Flux (
   -- * Dispatcher
   -- $dispatcher
@@ -155,18 +157,6 @@ reactRenderViewToString includeStatic (View rc) = do
   sRef <- (if includeStatic then js_ReactRenderStaticMarkup else js_ReactRenderToString) e
   mtxt <- fromJSVal sRef
   maybe (error "Unable to convert string return to Text") return mtxt
-
-foreign import javascript unsafe
-    "(typeof ReactDOM === 'object' ? ReactDOM : React)['render']($1, document.getElementById($2))"
-    js_ReactRender :: ReactElementRef -> JSString -> IO ()
-
-foreign import javascript unsafe
-    "(typeof ReactDOMServer === 'object' ? ReactDOMServer : (typeof ReactDOM === 'object' ? ReactDOM : React))['renderToString']($1)"
-    js_ReactRenderToString :: ReactElementRef -> IO JSVal
-
-foreign import javascript unsafe
-    "(typeof ReactDOMServer === 'object' ? ReactDOMServer : (typeof ReactDOM === 'object' ? ReactDOM : React))['renderToStaticMarkup']($1)"
-    js_ReactRenderStaticMarkup :: ReactElementRef -> IO JSVal
 
 
 -- $performance
@@ -312,3 +302,30 @@ foreign import javascript unsafe
 --
 -- >dispatchTodo :: TodoAction -> [SomeStoreAction]
 -- >dispatchTodo a = [SomeStoreAction todoStore a]
+
+#ifdef __GHCJS__
+
+foreign import javascript unsafe
+    "(typeof ReactDOM === 'object' ? ReactDOM : React)['render']($1, document.getElementById($2))"
+    js_ReactRender :: ReactElementRef -> JSString -> IO ()
+
+foreign import javascript unsafe
+    "(typeof ReactDOMServer === 'object' ? ReactDOMServer : (typeof ReactDOM === 'object' ? ReactDOM : React))['renderToString']($1)"
+    js_ReactRenderToString :: ReactElementRef -> IO JSVal
+
+foreign import javascript unsafe
+    "(typeof ReactDOMServer === 'object' ? ReactDOMServer : (typeof ReactDOM === 'object' ? ReactDOM : React))['renderToStaticMarkup']($1)"
+    js_ReactRenderStaticMarkup :: ReactElementRef -> IO JSVal
+
+#else
+
+js_ReactRender :: ReactElementRef -> JSString -> IO ()
+js_ReactRender _ _ = error "js_ReactRender only works with GHCJS"
+
+js_ReactRenderToString :: ReactElementRef -> IO JSVal
+js_ReactRenderToString _ = error "js_ReactRenderToString only works with GHCJS"
+
+js_ReactRenderStaticMarkup :: ReactElementRef -> IO JSVal
+js_ReactRenderStaticMarkup _ = error "js_ReactRenderStaticMarkup only works with GHCJS"
+
+#endif

--- a/react-hs/src/React/Flux/Addons/Bootstrap.hs
+++ b/react-hs/src/React/Flux/Addons/Bootstrap.hs
@@ -7,6 +7,9 @@
 -- together.  While I am leaving this here for backwards compatibility, I suggest you use a different
 -- library.  I am currently using <http://www.material-ui.com Material UI> and accessing the
 -- components using @foreign_@.
+
+{-# LANGUAGE CPP #-}
+
 module React.Flux.Addons.Bootstrap (
     bootstrap_
 ) where
@@ -40,6 +43,15 @@ bootstrap_ :: String
 
 bootstrap_ n = foreignClass (js_ReactBootstrap $ toJSString n)
 
+#ifdef __GHCJS__
+
 foreign import javascript unsafe
     "window['ReactBootstrap'][$1]"
     js_ReactBootstrap :: JSString -> JSVal
+
+#else
+
+js_ReactBootstrap :: JSString -> JSVal
+js_ReactBootstrap _ = error "js_ReactBootstrap only works with GHCJS"
+
+#endif

--- a/react-hs/src/React/Flux/Addons/Intl.hs
+++ b/react-hs/src/React/Flux/Addons/Intl.hs
@@ -80,7 +80,8 @@
 --        locale into a store and use a controller-view to pass the locale and lookup the messages
 --        for 'intlProvider_'.
 
-{-# LANGUAGE TemplateHaskell, QuasiQuotes, OverloadedStrings #-}
+{-# LANGUAGE CPP, TemplateHaskell, QuasiQuotes, OverloadedStrings #-}
+
 module React.Flux.Addons.Intl(
     intlProvider_
 
@@ -159,46 +160,6 @@ import qualified Data.Text.Encoding as T
 import GHCJS.Types (JSString, JSVal)
 import GHCJS.Marshal (ToJSVal(..))
 import qualified JavaScript.Object as JSO
-
-foreign import javascript unsafe
-    "$r = ReactIntl['IntlProvider']"
-    js_intlProvider :: JSVal
-
-foreign import javascript unsafe
-    "$r = ReactIntl['FormattedNumber']"
-    js_formatNumber :: JSVal
-
-foreign import javascript unsafe
-    "$r = ReactIntl['FormattedDate']"
-    js_formatDate :: JSVal
-
-foreign import javascript unsafe
-    "$r = ReactIntl['FormattedRelative']"
-    js_formatRelative :: JSVal
-
-foreign import javascript unsafe
-    "$r = ReactIntl['FormattedPlural']"
-    js_formatPlural :: JSVal
-
-foreign import javascript unsafe
-    "$r = ReactIntl['FormattedMessage']"
-    js_formatMsg :: JSVal
-
-foreign import javascript unsafe
-    "$r = ReactIntl['FormattedHTMLMessage']"
-    js_formatHtmlMsg :: JSVal
-
-foreign import javascript unsafe
-    "$r = (new Date($1, $2-1, $3))"
-    js_mkDate :: Int -> Int -> Int -> JSVal
-
-foreign import javascript unsafe
-    "$r = (new Date(Date.UTC($1, $2-1, $3, $4, $5, $6, $7)))"
-    js_mkDateTime :: Int -> Int -> Int -> Int -> Int -> Int -> Int -> JSVal
-
-foreign import javascript unsafe
-    "$1['intl'][$2]($3, $4)"
-    js_callContextAPI :: JSVal -> JSString -> JSVal -> JSO.Object -> IO JSVal
 
 -- | Convert a day to a javascript Date.  This is useful to pass a date as a property to a
 -- 'message'.  Note that @JSVal@ is an instance of @ToJSVal@ so the result of 'dayToJSVal' can
@@ -697,3 +658,79 @@ escapeForXml = T.concatMap f
         f '\'' = "&apos;"
         f x | isPrint x || x == '\n' = T.singleton x
         f x = "&#" <> T.pack (show $ ord x) <> ";"
+
+#ifdef __GHCJS__
+
+foreign import javascript unsafe
+    "$r = ReactIntl['IntlProvider']"
+    js_intlProvider :: JSVal
+
+foreign import javascript unsafe
+    "$r = ReactIntl['FormattedNumber']"
+    js_formatNumber :: JSVal
+
+foreign import javascript unsafe
+    "$r = ReactIntl['FormattedDate']"
+    js_formatDate :: JSVal
+
+foreign import javascript unsafe
+    "$r = ReactIntl['FormattedRelative']"
+    js_formatRelative :: JSVal
+
+foreign import javascript unsafe
+    "$r = ReactIntl['FormattedPlural']"
+    js_formatPlural :: JSVal
+
+foreign import javascript unsafe
+    "$r = ReactIntl['FormattedMessage']"
+    js_formatMsg :: JSVal
+
+foreign import javascript unsafe
+    "$r = ReactIntl['FormattedHTMLMessage']"
+    js_formatHtmlMsg :: JSVal
+
+foreign import javascript unsafe
+    "$r = (new Date($1, $2-1, $3))"
+    js_mkDate :: Int -> Int -> Int -> JSVal
+
+foreign import javascript unsafe
+    "$r = (new Date(Date.UTC($1, $2-1, $3, $4, $5, $6, $7)))"
+    js_mkDateTime :: Int -> Int -> Int -> Int -> Int -> Int -> Int -> JSVal
+
+foreign import javascript unsafe
+    "$1['intl'][$2]($3, $4)"
+    js_callContextAPI :: JSVal -> JSString -> JSVal -> JSO.Object -> IO JSVal
+
+#else
+
+js_intlProvider :: JSVal
+js_intlProvider = error "js_intlProvider only works with GHCJS"
+
+js_formatNumber :: JSVal
+js_formatNumber = error "js_formatNumber only works with GHCJS"
+
+js_formatDate :: JSVal
+js_formatDate = error "js_formatDate only works with GHCJS"
+
+js_formatRelative :: JSVal
+js_formatRelative = error "js_formatRelative only works with GHCJS"
+
+js_formatPlural :: JSVal
+js_formatPlural = error "js_formatPlural only works with GHCJS"
+
+js_formatMsg :: JSVal
+js_formatMsg = error "js_formatMsg only works with GHCJS"
+
+js_formatHtmlMsg :: JSVal
+js_formatHtmlMsg = error "js_formatHtmlMsg only works with GHCJS"
+
+js_mkDate :: Int -> Int -> Int -> JSVal
+js_mkDate _ _ _ = error "js_mkDate only works with GHCJS"
+
+js_mkDateTime :: Int -> Int -> Int -> Int -> Int -> Int -> Int -> JSVal
+js_mkDateTime _ _ _ _ _ _ _ = error "js_mkDateTime only works with GHCJS"
+
+js_callContextAPI :: JSVal -> JSString -> JSVal -> JSO.Object -> IO JSVal
+js_callContextAPI _ _ _ _ = error "js_callContextAPI only works with GHCJS"
+
+#endif

--- a/react-hs/src/React/Flux/Addons/React.hs
+++ b/react-hs/src/React/Flux/Addons/React.hs
@@ -1,4 +1,4 @@
-{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE CPP, TypeApplications #-}
 
 -- | Bindings for the <https://facebook.github.io/react/docs/addons.html React addons> that make
 -- sense to use from Haskell.  At the moment, that is only the
@@ -27,12 +27,7 @@ import GHCJS.Types (JSVal, JSString)
 -- >cssTransitionGroup ["transitionName" $= "example", transitionAppear @= True, transitionAppearTimeout @= (100 :: Int)] $
 -- >    h1_ "Fading at initial mount"
 cssTransitionGroup :: [PropertyOrHandler eventHandler] -> ReactElementM eventHandler a -> ReactElementM eventHandler a
-
 cssTransitionGroup props children = foreignClass js_CSSTransitionGroup props children
-
-foreign import javascript unsafe
-    "React['addons']['CSSTransitionGroup']"
-    js_CSSTransitionGroup :: JSVal
 
 
 --------------------------------------------------------------------------------
@@ -103,6 +98,22 @@ perfToggleButton = mkControllerView @'[StoreArg PerfStoreData] "perf toggle butt
 perfToggleButton_ :: [PerfPrint] -> ReactElementM handler ()
 perfToggleButton_ toPrint = view_ perfToggleButton "perf-toggle-button" toPrint
 
+#ifdef __GHCJS__
+
+foreign import javascript unsafe
+    "React['addons']['CSSTransitionGroup']"
+    js_CSSTransitionGroup :: JSVal
+
 foreign import javascript unsafe
     "React['addons']['Perf'][$1]()"
     js_perf :: JSString -> IO ()
+
+#else
+
+js_CSSTransitionGroup :: JSVal
+js_CSSTransitionGroup = error "js_CSSTransitionGroup only works with GHCJS"
+
+js_perf :: JSString -> IO ()
+js_perf _ = error "js_perf only works with GHCJS"
+
+#endif

--- a/react-hs/src/React/Flux/Ajax.hs
+++ b/react-hs/src/React/Flux/Ajax.hs
@@ -2,6 +2,9 @@
 -- in that it mostly directly exposes the XMLHttpRequest access.  If you are using
 -- servant, <http://hackage.haskell.org/package/react-flux-servant react-flux-servant>
 -- for a higher-level interface.
+
+{-# LANGUAGE CPP #-}
+
 module React.Flux.Ajax (
     initAjax
   , RequestTimeout(..)
@@ -180,6 +183,10 @@ jsonAjax timeout method uri headers body handler = do
                     Right v  -> handler $ Right v
             else handler $ Left (respStatus resp, JSS.textFromJSString $ respResponseText resp)
 
+instance ToJSVal AjaxRequest
+
+#ifdef __GHCJS__
+
 foreign import javascript unsafe
     "hsreact$ajax($1, $2)"
     js_ajax :: JSVal -> Export HandlerWrapper -> IO ()
@@ -208,4 +215,27 @@ foreign import javascript unsafe
     "h$release($1)"
     js_release :: Export HandlerWrapper -> IO ()
 
-instance ToJSVal AjaxRequest
+#else
+
+js_ajax :: JSVal -> Export HandlerWrapper -> IO ()
+js_ajax _ _ = error "js_ajax only works with GHCJS"
+
+js_setAjaxCallback :: Callback (JSVal -> JSVal -> IO ()) -> IO ()
+js_setAjaxCallback _ = error "js_setAjaxCallback only works with GHCJS"
+
+js_getHandlerWrapper :: JSVal -> IO (Export HandlerWrapper)
+js_getHandlerWrapper _ = error "js_getHandlerWrapper only works with GHCJS"
+
+js_reqTimedOut :: JSVal -> Bool
+js_reqTimedOut _ = error "js_reqTimedOut only works with GHCJS"
+
+js_xhrStatus :: JSVal -> IO Int
+js_xhrStatus _ = error "js_xhrStatus only works with GHCJS"
+
+js_xhrResponseText :: JSVal -> IO JSString
+js_xhrResponseText _ = error "js_xhrResponseText only works with GHCJS"
+
+js_release :: Export HandlerWrapper -> IO ()
+js_release _ = error "js_release only works with GHCJS"
+
+#endif

--- a/react-hs/src/React/Flux/Combinators.hs
+++ b/react-hs/src/React/Flux/Combinators.hs
@@ -1,7 +1,7 @@
 -- | This module contains some useful combinators I have come across as I built a large
 -- react-flux application.  None of these are required to use React.Flux, they just reduce somewhat
 -- the typing needed to create rendering functions.
-{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE CPP, DeriveAnyClass #-}
 module React.Flux.Combinators (
     clbutton_
   , cldiv_
@@ -16,10 +16,6 @@ import React.Flux.DOM
 import React.Flux.Internal
 import React.Flux.PropertiesAndEvents
 import GHCJS.Types (JSVal)
-
-foreign import javascript unsafe
-    "$r = window[$1]"
-    js_lookupWindow :: JSString -> JSVal
 
 -- | A wrapper around 'foreignClass' that looks up the class on the `window`.  I use it for several
 -- third-party react components.  For example, with <https://github.com/reactjs/react-modal react-modal>,
@@ -106,3 +102,16 @@ faIcon_ cl = i_ ["className" &= ("fa fa-" <> cl)] mempty
 -- on elements.
 style :: [(JSString,JSString)] -> PropertyOrHandler handler
 style = nestedProperty "style" . map (\(n,a) -> (n &= a))
+
+#ifdef __GHCJS__
+
+foreign import javascript unsafe
+    "$r = window[$1]"
+    js_lookupWindow :: JSString -> JSVal
+
+#else
+
+js_lookupWindow :: JSString -> JSVal
+js_lookupWindow _ = error "js_lookupWindow only works with GHCJS"
+
+#endif

--- a/react-hs/src/React/Flux/Internal.hs
+++ b/react-hs/src/React/Flux/Internal.hs
@@ -1,4 +1,4 @@
-{-# LANGUAGE ViewPatterns #-}
+{-# LANGUAGE CPP, ViewPatterns #-}
 {-# OPTIONS_GHC -fno-warn-redundant-constraints #-}
 
 -- | Internal module for React.Flux
@@ -400,66 +400,6 @@ mkReactElement runHandler this = runWriterT . mToElem
 
 type MkReactElementM a = WriterT [CallbackToRelease] IO a
 
-foreign import javascript unsafe
-    "React['createElement']($1)"
-    js_ReactCreateElementNoChildren :: JSVal -> IO ReactElementRef
-
-foreign import javascript unsafe
-    "React['createElement']($1, $2, $3)"
-    js_ReactCreateElementName :: JSString -> JSO.Object -> JSVal -> IO ReactElementRef
-
-foreign import javascript unsafe
-    "React['createElement']($1, $2, $3)"
-    js_ReactCreateForeignElement :: ReactViewRef a -> JSO.Object -> JSVal -> IO ReactElementRef
-
-foreign import javascript unsafe
-    "React['createElement']($1, {hs:$2}, $3)"
-    js_ReactCreateClass :: ReactViewRef a -> Export props -> JSVal -> IO ReactElementRef
-
-foreign import javascript unsafe
-    "React['createElement']($1, {key: $2, hs:$3}, $4)"
-    js_ReactCreateKeyedElement :: ReactViewRef a -> JSVal -> Export props -> JSVal -> IO ReactElementRef
-
-foreign import javascript unsafe
-    "React['createElement']($1, {key: $2, hs:$3})"
-    js_ReactNewViewElement :: ReactViewRef a -> JSVal -> NewJsProps -> IO ReactElementRef
-
-foreign import javascript unsafe
-    "React['createElement']($1, {hs:$2})"
-    js_ReactNewViewElementNoKey :: ReactViewRef a -> NewJsProps -> IO ReactElementRef
-
-foreign import javascript unsafe
-    "hsreact$mk_arguments_callback($1)"
-    js_CreateArgumentsCallback :: Callback (JSVal -> IO ()) -> IO JSVal
-
-foreign import javascript unsafe
-    "hsreact$wrap_callback_returning_element($1)"
-    js_wrapCallbackReturningElement :: Callback (JSVal -> JSVal -> IO ()) -> IO JSVal
-
-foreign import javascript unsafe
-    "$1.elem = $2"
-    js_setElemReturnFromCallback :: JSVal -> ReactElementRef -> IO ()
-
-foreign import javascript unsafe
-    "$r = hsreact$divLikeElement"
-    js_divLikeElement :: JSVal
-
-foreign import javascript unsafe
-    "$r = hsreact$textWrapper"
-    js_textWrapper :: JSVal
-
-foreign import javascript unsafe
-    "$1['context']"
-    js_ReactGetContext :: ReactThis state props -> IO JSVal
-
-foreign import javascript unsafe
-    "hsreact$children_to_array($1['props']['children'])"
-    js_ReactGetChildren :: ReactThis state props -> IO JSArray
-
-foreign import javascript unsafe
-    "[]"
-    js_emptyList :: IO NewJsProps
-
 js_ReactCreateContent :: JSString -> ReactElementRef
 js_ReactCreateContent = ReactElementRef . unsafeCoerce
 
@@ -607,8 +547,122 @@ allEq__ Proxy i jsas jsbs = do
   (&&) <$> singleEq_ (Proxy :: Proxy t)  jsa jsb
        <*> allEq_    (Proxy :: Proxy ts) (i + 1) jsas jsbs
 
+#ifdef __GHCJS__
+
+foreign import javascript unsafe
+    "React['createElement']($1)"
+    js_ReactCreateElementNoChildren :: JSVal -> IO ReactElementRef
+
+foreign import javascript unsafe
+    "React['createElement']($1, $2, $3)"
+    js_ReactCreateElementName :: JSString -> JSO.Object -> JSVal -> IO ReactElementRef
+
+foreign import javascript unsafe
+    "React['createElement']($1, $2, $3)"
+    js_ReactCreateForeignElement :: ReactViewRef a -> JSO.Object -> JSVal -> IO ReactElementRef
+
+foreign import javascript unsafe
+    "React['createElement']($1, {hs:$2}, $3)"
+    js_ReactCreateClass :: ReactViewRef a -> Export props -> JSVal -> IO ReactElementRef
+
+foreign import javascript unsafe
+    "React['createElement']($1, {key: $2, hs:$3}, $4)"
+    js_ReactCreateKeyedElement :: ReactViewRef a -> JSVal -> Export props -> JSVal -> IO ReactElementRef
+
+foreign import javascript unsafe
+    "React['createElement']($1, {key: $2, hs:$3})"
+    js_ReactNewViewElement :: ReactViewRef a -> JSVal -> NewJsProps -> IO ReactElementRef
+
+foreign import javascript unsafe
+    "React['createElement']($1, {hs:$2})"
+    js_ReactNewViewElementNoKey :: ReactViewRef a -> NewJsProps -> IO ReactElementRef
+
+foreign import javascript unsafe
+    "hsreact$mk_arguments_callback($1)"
+    js_CreateArgumentsCallback :: Callback (JSVal -> IO ()) -> IO JSVal
+
+foreign import javascript unsafe
+    "hsreact$wrap_callback_returning_element($1)"
+    js_wrapCallbackReturningElement :: Callback (JSVal -> JSVal -> IO ()) -> IO JSVal
+
+foreign import javascript unsafe
+    "$1.elem = $2"
+    js_setElemReturnFromCallback :: JSVal -> ReactElementRef -> IO ()
+
+foreign import javascript unsafe
+    "$r = hsreact$divLikeElement"
+    js_divLikeElement :: JSVal
+
+foreign import javascript unsafe
+    "$r = hsreact$textWrapper"
+    js_textWrapper :: JSVal
+
+foreign import javascript unsafe
+    "$1['context']"
+    js_ReactGetContext :: ReactThis state props -> IO JSVal
+
+foreign import javascript unsafe
+    "hsreact$children_to_array($1['props']['children'])"
+    js_ReactGetChildren :: ReactThis state props -> IO JSArray
+
+foreign import javascript unsafe
+    "[]"
+    js_emptyList :: IO NewJsProps
+
 -- | (This is similar to findFromState, but less specific, and more "pure".  not sure if we can merge
 -- the two?)
 foreign import javascript unsafe
   "$2[$1]"
   js_findFromArray :: Int -> JSVal -> JSVal
+
+#else
+
+js_ReactCreateElementNoChildren :: JSVal -> IO ReactElementRef
+js_ReactCreateElementNoChildren _ = error "js_ReactCreateElementNoChildren only works with GHCJS"
+
+js_ReactCreateElementName :: JSString -> JSO.Object -> JSVal -> IO ReactElementRef
+js_ReactCreateElementName _ _ _ = error "js_ReactCreateElementName only works with GHCJS"
+
+js_ReactCreateForeignElement :: ReactViewRef a -> JSO.Object -> JSVal -> IO ReactElementRef
+js_ReactCreateForeignElement _ _ _ = error "js_ReactCreateForeignElement only works with GHCJS"
+
+js_ReactCreateClass :: ReactViewRef a -> Export props -> JSVal -> IO ReactElementRef
+js_ReactCreateClass _ _ _ = error "js_ReactCreateClass only works with GHCJS"
+
+js_ReactCreateKeyedElement :: ReactViewRef a -> JSVal -> Export props -> JSVal -> IO ReactElementRef
+js_ReactCreateKeyedElement _ _ _ _ = error "js_ReactCreateKeyedElement only works with GHCJS"
+
+js_ReactNewViewElement :: ReactViewRef a -> JSVal -> NewJsProps -> IO ReactElementRef
+js_ReactNewViewElement _ _ _ = error "js_ReactNewViewElement only works with GHCJS"
+
+js_ReactNewViewElementNoKey :: ReactViewRef a -> NewJsProps -> IO ReactElementRef
+js_ReactNewViewElementNoKey _ _ = error "js_ReactNewViewElementNoKey only works with GHCJS"
+
+js_CreateArgumentsCallback :: Callback (JSVal -> IO ()) -> IO JSVal
+js_CreateArgumentsCallback _ = error "js_CreateArgumentsCallback only works with GHCJS"
+
+js_wrapCallbackReturningElement :: Callback (JSVal -> JSVal -> IO ()) -> IO JSVal
+js_wrapCallbackReturningElement _ = error "js_wrapCallbackReturningElement only works with GHCJS"
+
+js_setElemReturnFromCallback :: JSVal -> ReactElementRef -> IO ()
+js_setElemReturnFromCallback _ _ = error "js_setElemReturnFromCallback only works with GHCJS"
+
+js_divLikeElement :: JSVal
+js_divLikeElement = error "js_divLikeElement only works with GHCJS"
+
+js_textWrapper :: JSVal
+js_textWrapper = error "js_textWrapper only works with GHCJS"
+
+js_ReactGetContext :: ReactThis state props -> IO JSVal
+js_ReactGetContext _ = error "js_ReactGetContext only works with GHCJS"
+
+js_ReactGetChildren :: ReactThis state props -> IO JSArray
+js_ReactGetChildren _ = error "js_ReactGetChildren only works with GHCJS"
+
+js_emptyList :: IO NewJsProps
+js_emptyList = error "js_emptyList only works with GHCJS"
+
+js_findFromArray :: Int -> JSVal -> JSVal
+js_findFromArray _ _ = error "js_findFromArray only works with GHCJS"
+
+#endif

--- a/react-hs/src/React/Flux/Outdated.hs
+++ b/react-hs/src/React/Flux/Outdated.hs
@@ -1,5 +1,5 @@
 -- | Internal module containing the view definitions
-{-# LANGUAGE UndecidableInstances, TypeApplications, MagicHash #-}
+{-# LANGUAGE CPP, UndecidableInstances, TypeApplications, MagicHash #-}
 {-# OPTIONS_GHC -fno-warn-redundant-constraints #-}
 module React.Flux.Outdated
   ( SomeStoreAction(..)
@@ -89,32 +89,8 @@ runStateViewHandler this handler = do
 --- Various GHCJS only utilities
 ---------------------------------------------------------------------------------------------------
 
-foreign import javascript unsafe
-    "$1['state'].hs"
-    js_ReactGetState :: ReactThis state props -> IO (Export state)
-
-foreign import javascript unsafe
-    "$1['props'].hs"
-    js_ReactGetProps :: ReactThis state props -> IO (Export props)
-
-foreign import javascript unsafe
-    "$1._updateAndReleaseState($2)"
-    js_ReactUpdateAndReleaseState :: ReactThis state props -> Export state -> IO ()
-
 newtype RenderCbArg = RenderCbArg JSVal
 instance IsJSVal RenderCbArg
-
-foreign import javascript unsafe
-    "$1.newCallbacks = $2; $1.elem = $3;"
-    js_RenderCbSetResults :: RenderCbArg -> JSVal -> ReactElementRef -> IO ()
-
-foreign import javascript unsafe
-    "hsreact$mk_lifecycle_view($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)"
-    js_makeLifecycleView :: JSString -> Export state -> Callback (JSVal -> JSVal -> IO ())
-                         -> JSVal -> JSVal -> JSVal -> JSVal -> JSVal -> JSVal
-                         -> Callback (JSVal -> JSVal -> IO JSVal)
-                         -> Callback (JSVal -> JSVal -> IO JSVal)
-                         -> IO (ReactViewRef props)
 
 mkRenderCallback :: Typeable props
                  => (ReactThis state props -> IO state) -- ^ parse state
@@ -466,17 +442,6 @@ mkLCallback2 (Just f) c = do
     c (f ps) this argRef
   return $ jsval cb
 
--- React 0.13 has React.findDOMNode, while 0.14 moves it to ReactDOM.findDOMNode.  Also, 0.14
--- does not need to call findDOMNode on refs.
-
-foreign import javascript unsafe
-    "typeof ReactDOM === 'object' ? ReactDOM['findDOMNode']($1) : React['findDOMNode']($1)"
-    js_ReactFindDOMNode :: ReactThis state props -> IO JSVal
-
-foreign import javascript unsafe
-    "typeof ReactDOM === 'object' ? $1['refs'][$2] : React['findDOMNode']($1['refs'][$2])"
-    js_ReactGetRef :: ReactThis state props -> JSString -> IO JSVal
-
 
 ----------------------------------------------------------------------------------------------------
 -- reactRender has two versions
@@ -494,10 +459,6 @@ reactRender :: Typeable props
 reactRender htmlId rc props = do
     (e, _) <- mkReactElement id (ReactThis nullRef) $ view rc props mempty
     js_ReactRender e (toJSString htmlId)
-
-foreign import javascript unsafe
-    "(typeof ReactDOM === 'object' ? ReactDOM : React)['render']($1, document.getElementById($2))"
-    js_ReactRender :: ReactElementRef -> JSString -> IO ()
 
 -- | Render your React application to a string using either @ReactDOMServer.renderToString@ if the first
 -- argument is false or @ReactDOMServer.renderToStaticMarkup@ if the first argument is true.
@@ -522,6 +483,46 @@ reactRenderToString includeStatic rc props = do
     mtxt <- fromJSVal sRef
     maybe (error "Unable to convert string return to Text") return mtxt
 
+#ifdef __GHCJS__
+
+foreign import javascript unsafe
+    "$1['state'].hs"
+    js_ReactGetState :: ReactThis state props -> IO (Export state)
+
+foreign import javascript unsafe
+    "$1['props'].hs"
+    js_ReactGetProps :: ReactThis state props -> IO (Export props)
+
+foreign import javascript unsafe
+    "$1._updateAndReleaseState($2)"
+    js_ReactUpdateAndReleaseState :: ReactThis state props -> Export state -> IO ()
+
+foreign import javascript unsafe
+    "$1.newCallbacks = $2; $1.elem = $3;"
+    js_RenderCbSetResults :: RenderCbArg -> JSVal -> ReactElementRef -> IO ()
+
+foreign import javascript unsafe
+    "hsreact$mk_lifecycle_view($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)"
+    js_makeLifecycleView :: JSString -> Export state -> Callback (JSVal -> JSVal -> IO ())
+                         -> JSVal -> JSVal -> JSVal -> JSVal -> JSVal -> JSVal
+                         -> Callback (JSVal -> JSVal -> IO JSVal)
+                         -> Callback (JSVal -> JSVal -> IO JSVal)
+                         -> IO (ReactViewRef props)
+
+-- React 0.13 has React.findDOMNode, while 0.14 moves it to ReactDOM.findDOMNode.  Also, 0.14
+-- does not need to call findDOMNode on refs.
+foreign import javascript unsafe
+    "typeof ReactDOM === 'object' ? ReactDOM['findDOMNode']($1) : React['findDOMNode']($1)"
+    js_ReactFindDOMNode :: ReactThis state props -> IO JSVal
+
+foreign import javascript unsafe
+    "typeof ReactDOM === 'object' ? $1['refs'][$2] : React['findDOMNode']($1['refs'][$2])"
+    js_ReactGetRef :: ReactThis state props -> JSString -> IO JSVal
+
+foreign import javascript unsafe
+    "(typeof ReactDOM === 'object' ? ReactDOM : React)['render']($1, document.getElementById($2))"
+    js_ReactRender :: ReactElementRef -> JSString -> IO ()
+
 foreign import javascript unsafe
     "(typeof ReactDOMServer === 'object' ? ReactDOMServer : (typeof ReactDOM === 'object' ? ReactDOM : React))['renderToString']($1)"
     js_ReactRenderToString :: ReactElementRef -> IO JSVal
@@ -529,3 +530,43 @@ foreign import javascript unsafe
 foreign import javascript unsafe
     "(typeof ReactDOMServer === 'object' ? ReactDOMServer : (typeof ReactDOM === 'object' ? ReactDOM : React))['renderToStaticMarkup']($1)"
     js_ReactRenderStaticMarkup :: ReactElementRef -> IO JSVal
+
+#else
+
+js_ReactGetState :: ReactThis state props -> IO (Export state)
+js_ReactGetState _ = error "js_ReactGetState only works with GHCJS"
+
+js_ReactGetProps :: ReactThis state props -> IO (Export props)
+js_ReactGetProps _ = error "js_ReactGetProps only works with GHCJS"
+
+js_ReactUpdateAndReleaseState :: ReactThis state props -> Export state -> IO ()
+js_ReactUpdateAndReleaseState _ _ = error "js_ReactUpdateAndReleaseState only works with GHCJS"
+
+js_RenderCbSetResults :: RenderCbArg -> JSVal -> ReactElementRef -> IO ()
+js_RenderCbSetResults _ _ _ = error "js_RenderCbSetResults only works with GHCJS"
+
+js_makeLifecycleView :: JSString -> Export state -> Callback (JSVal -> JSVal -> IO ())
+                     -> JSVal -> JSVal -> JSVal -> JSVal -> JSVal -> JSVal
+                     -> Callback (JSVal -> JSVal -> IO JSVal)
+                     -> Callback (JSVal -> JSVal -> IO JSVal)
+                     -> IO (ReactViewRef props)
+js_makeLifecycleView _ _ _ _ _ _ _ _ _ _ _ = error "js_makeLifecycleView only works with GHCJS"
+
+-- React 0.13 has React.findDOMNode, while 0.14 moves it to ReactDOM.findDOMNode.  Also, 0.14
+-- does not need to call findDOMNode on refs.
+js_ReactFindDOMNode :: ReactThis state props -> IO JSVal
+js_ReactFindDOMNode _ = error "js_ReactFindDOMNode only works with GHCJS"
+
+js_ReactGetRef :: ReactThis state props -> JSString -> IO JSVal
+js_ReactGetRef _ _ = error "js_ReactGetRef only works with GHCJS"
+
+js_ReactRender :: ReactElementRef -> JSString -> IO ()
+js_ReactRender _ _ = error "js_ReactRender only works with GHCJS"
+
+js_ReactRenderToString :: ReactElementRef -> IO JSVal
+js_ReactRenderToString _ = error "js_ReactRenderToString only works with GHCJS"
+
+js_ReactRenderStaticMarkup :: ReactElementRef -> IO JSVal
+js_ReactRenderStaticMarkup _ = error "js_ReactRenderStaticMarkup only works with GHCJS"
+
+#endif

--- a/react-hs/src/React/Flux/PropertiesAndEvents.hs
+++ b/react-hs/src/React/Flux/PropertiesAndEvents.hs
@@ -1,7 +1,7 @@
 -- | This module contains the definitions for creating properties to pass to javascript elements and
 -- foreign javascript classes.  In addition, it contains definitions for the
 -- <https://facebook.github.io/react/docs/events.html React Event System>.
-{-# LANGUAGE ViewPatterns, UndecidableInstances #-}
+{-# LANGUAGE CPP, ViewPatterns, UndecidableInstances #-}
 module React.Flux.PropertiesAndEvents (
     PropertyOrHandler
 
@@ -299,14 +299,6 @@ preventDefault (evtHandlerArg -> HandlerArg ref) = unsafePerformIO (js_preventDe
 stopPropagation :: Event -> ()
 stopPropagation (evtHandlerArg -> HandlerArg ref) = unsafePerformIO (js_stopPropagation ref) `seq` ()
 
-foreign import javascript unsafe
-    "$1['preventDefault']();"
-    js_preventDefault :: JSVal -> IO ()
-
-foreign import javascript unsafe
-    "$1['stopPropagation']();"
-    js_stopPropagation :: JSVal -> IO ()
-
 -- | By default, the handlers below are triggered during the bubbling phase.  Use this to switch
 -- them to trigger during the capture phase.
 capturePhase :: PropertyOrHandler handler -> PropertyOrHandler handler
@@ -500,11 +492,6 @@ onMouseUp = on2 "onMouseUp" parseMouseEvent
 -- Touch
 --------------------------------------------------------------------------------
 
--- | Initialize touch events is only needed with React 0.13, in version 0.14 it was removed.
-foreign import javascript unsafe
-    "React['initializeTouchEvents'] ? React['initializeTouchEvents'](true) : null"
-    initializeTouchEvents :: IO ()
-
 data Touch = Touch {
     touchIdentifier :: Int
   , touchTarget :: EventTarget
@@ -627,23 +614,11 @@ onError = on "onError"
 --- JS Utils
 --------------------------------------------------------------------------------
 
-foreign import javascript unsafe
-    "$1[$2]"
-    js_getProp :: JSVal -> JSString -> JSVal
-
-foreign import javascript unsafe
-    "$1[$2]"
-    js_getArrayProp :: JSVal -> JSString -> JSArray
-
 -- | Access a property from an object.  Since event objects are immutable, we can use
 -- unsafePerformIO without worry.
 (.:) :: FromJSVal b => JSVal -> JSString -> b
 obj .: key = fromMaybe (error "Unable to decode event target") $ unsafePerformIO $  -- TODO: get rid of the unsafePerformIO here!
     fromJSVal $ js_getProp obj key
-
-foreign import javascript unsafe
-    "$1['getModifierState']($2)"
-    js_GetModifierState :: JSVal -> JSString -> JSVal
 
 getModifierState :: JSVal -> T.Text -> Bool
 getModifierState ref = fromJSBool . js_GetModifierState ref . JSS.textToJSString
@@ -653,3 +628,52 @@ arrayLength = JSA.length
 
 arrayIndex :: Int -> JSArray -> JSVal
 arrayIndex = JSA.index
+
+#ifdef __GHCJS__
+
+foreign import javascript unsafe
+    "$1['preventDefault']();"
+    js_preventDefault :: JSVal -> IO ()
+
+foreign import javascript unsafe
+    "$1['stopPropagation']();"
+    js_stopPropagation :: JSVal -> IO ()
+
+-- | Initialize touch events is only needed with React 0.13, in version 0.14 it was removed.
+foreign import javascript unsafe
+    "React['initializeTouchEvents'] ? React['initializeTouchEvents'](true) : null"
+    initializeTouchEvents :: IO ()
+
+foreign import javascript unsafe
+    "$1[$2]"
+    js_getProp :: JSVal -> JSString -> JSVal
+
+foreign import javascript unsafe
+    "$1[$2]"
+    js_getArrayProp :: JSVal -> JSString -> JSArray
+
+foreign import javascript unsafe
+    "$1['getModifierState']($2)"
+    js_GetModifierState :: JSVal -> JSString -> JSVal
+
+#else
+
+js_preventDefault :: JSVal -> IO ()
+js_preventDefault _ = error "js_preventDefault only works with GHCJS"
+
+js_stopPropagation :: JSVal -> IO ()
+js_stopPropagation _ = error "js_stopPropagation only works with GHCJS"
+
+initializeTouchEvents :: IO ()
+initializeTouchEvents = error "initializeTouchEvents only works with GHCJS"
+
+js_getProp :: JSVal -> JSString -> JSVal
+js_getProp _ _ = error "js_getProp only works with GHCJS"
+
+js_getArrayProp :: JSVal -> JSString -> JSArray
+js_getArrayProp _ _ = error "js_getArrayProp only works with GHCJS"
+
+js_GetModifierState :: JSVal -> JSString -> JSVal
+js_GetModifierState _ _ = error "js_GetModifierState only works with GHCJS"
+
+#endif

--- a/react-hs/src/React/Flux/Store.hs
+++ b/react-hs/src/React/Flux/Store.hs
@@ -1,5 +1,5 @@
 -- | Internal module containing the store definitions.
-{-# LANGUAGE AllowAmbiguousTypes, OverloadedStrings #-}
+{-# LANGUAGE AllowAmbiguousTypes, CPP, OverloadedStrings #-}
 {-# OPTIONS_GHC -fno-warn-redundant-constraints #-}
 module React.Flux.Store (
     ReactStoreRef(..)
@@ -29,33 +29,12 @@ import Data.Monoid ((<>))
 import GHCJS.Foreign.Export
 import GHCJS.Types (JSVal, IsJSVal, JSString)
 import GHC.Fingerprint.Type
-import qualified Data.JSString.Int as JSString (decimal)
 import qualified Data.JSString as JSString
 import Data.Word (Word64)
 
-
--- | See https://github.com/ghcjs/ghcjs/issues/570 for details.
-decimal_workaround_570 :: Word64 -> JSString
-decimal_workaround_570 w = dropleadingzeros . mconcat $ showpadded <$> chunks
-  where
-    n :: Integer -> Integer
-    n i = 10^(5 * i)
-
-    chunks :: [Integer]
-    chunks =
-      [ (fromIntegral w `div` (n 3)) `mod` n 1
-      , (fromIntegral w `div` (n 2)) `mod` n 1
-      , (fromIntegral w `div` (n 1)) `mod` n 1
-      , fromIntegral w               `mod` n 1
-      ]
-
-    showpadded :: Integer -> JSString
-    showpadded i = JSString.reverse . JSString.take 5 . JSString.reverse
-                 $ JSString.pack "00000" <> JSString.decimal i
-
-    dropleadingzeros :: JSString -> JSString
-    dropleadingzeros = JSString.dropWhile (== '0')
-
+#ifdef __GHCJS__
+import qualified Data.JSString.Int as JSString (decimal)
+#endif
 
 -- | A store contains application state, receives actions from the dispatcher, and notifies
 -- controller-views to re-render themselves.  You can have multiple stores; it should be the case
@@ -211,6 +190,35 @@ readStoreData = do
   store :: NewReactStore storeData <- js_getNewStore (storeJsKey (Proxy :: Proxy storeData))
   js_getNewStoreData store >>= unsafeDerefExport "readStoreData"
 
+getNewStoreHS :: NewReactStore storeData -> IO NewReactStoreHS
+getNewStoreHS s = js_getNewStoreHS s >>= unsafeDerefExport "getNewStoreHS"
+{-# NOINLINE getNewStoreHS #-}
+
+#ifdef __GHCJS__
+
+-- | See https://github.com/ghcjs/ghcjs/issues/570 for details.
+decimal_workaround_570 :: Word64 -> JSString
+decimal_workaround_570 w = dropleadingzeros . mconcat $ showpadded <$> chunks
+  where
+    n :: Integer -> Integer
+    n i = 10^(5 * i)
+
+    chunks :: [Integer]
+    chunks =
+      [ (fromIntegral w `div` (n 3)) `mod` n 1
+      , (fromIntegral w `div` (n 2)) `mod` n 1
+      , (fromIntegral w `div` (n 1)) `mod` n 1
+      , fromIntegral w               `mod` n 1
+      ]
+
+    showpadded :: Integer -> JSString
+    showpadded i = JSString.reverse . JSString.take 5 . JSString.reverse
+                 $ JSString.pack "00000" <> JSString.decimal i
+
+
+    dropleadingzeros :: JSString -> JSString
+    dropleadingzeros = JSString.dropWhile (== '0')
+
 foreign import javascript unsafe
   "hsreact$storedata[$1]"
   js_getNewStore :: JSString -> IO (NewReactStore storeData)
@@ -218,10 +226,6 @@ foreign import javascript unsafe
 foreign import javascript unsafe
   "$1.hs"
   js_getNewStoreHS :: NewReactStore storeData -> IO (Export NewReactStoreHS)
-
-getNewStoreHS :: NewReactStore storeData -> IO NewReactStoreHS
-getNewStoreHS s = js_getNewStoreHS s >>= unsafeDerefExport "getNewStoreHS"
-{-# NOINLINE getNewStoreHS #-}
 
 foreign import javascript unsafe
   "$1.sdata"
@@ -235,3 +239,25 @@ foreign import javascript unsafe
 foreign import javascript unsafe
   "hsreact$transform_new_store($1, $2)"
   js_updateNewStore :: NewReactStore storeData -> Export storeData -> IO ()
+
+#else
+
+decimal_workaround_570 :: Word64 -> JSString
+decimal_workaround_570 _ = error "decimal_workaround_570 only works with GHCJS"
+
+js_getNewStore :: JSString -> IO (NewReactStore storeData)
+js_getNewStore _ = error "js_getNewStore only works with GHCJS"
+
+js_getNewStoreHS :: NewReactStore storeData -> IO (Export NewReactStoreHS)
+js_getNewStoreHS _ = error "js_getNewStoreHS only works with GHCJS"
+
+js_getNewStoreData :: NewReactStore storeData -> IO (Export storeData)
+js_getNewStoreData _ = error "js_getNewStoreData only works with GHCJS"
+
+js_createNewStore :: JSString -> Export storeData -> Export NewReactStoreHS -> IO ()
+js_createNewStore _ _ _ = error "js_createNewStore only works with GHCJS"
+
+js_updateNewStore :: NewReactStore storeData -> Export storeData -> IO ()
+js_updateNewStore _ _ = error "js_updateNewStore only works with GHCJS"
+
+#endif

--- a/react-hs/src/React/Flux/Views.hs
+++ b/react-hs/src/React/Flux/Views.hs
@@ -1,5 +1,5 @@
 -- | Internal module containing the view definitions
-{-# LANGUAGE UndecidableInstances, AllowAmbiguousTypes, TypeApplications, BangPatterns, ScopedTypeVariables #-}
+{-# LANGUAGE UndecidableInstances, AllowAmbiguousTypes, CPP, TypeApplications, BangPatterns, ScopedTypeVariables #-}
 {-# OPTIONS_GHC -fno-warn-redundant-constraints #-}
 module React.Flux.Views
   ( View(..)
@@ -181,14 +181,6 @@ getStoreJs :: Typeable store => JSVal -> IO store
 getStoreJs arg = js_getDeriveInput arg >>= unsafeDerefExport "getStoreJs"
 {-# NOINLINE getStoreJs #-} -- if this is inlined, GHCJS does not properly compile the getField callback
 
-foreign import javascript unsafe
-  "$1.input"
-  js_getDeriveInput :: JSVal -> IO (Export store)
-
-foreign import javascript unsafe
-  "$1.output = $2"
-  js_setDeriveOutput :: JSVal -> Export a -> IO ()
-
 
 ---------------------------------------------------------------------------------
 -- Public API Implementation
@@ -213,14 +205,6 @@ mkView name buildNode = unsafePerformIO $ do
   View <$> js_createNewView name renderCb propsEq
 {-# NOINLINE mkView #-}
 
-foreign import javascript unsafe
-    "hsreact$mk_new_view($1, $2, $3)"
-    js_createNewView
-        :: JSString
-        -> Callback (JSVal -> JSVal -> IO ())
-        -> Callback (JSVal -> JSVal -> IO JSVal)
-        -> IO (ReactViewRef props)
-
 mkStatefulView :: forall (state :: *) (props :: [*]).
                   (Typeable state, NFData state, Typeable state, Eq state,
                    ViewProps props, Typeable props, AllEq props)
@@ -244,16 +228,6 @@ mkStatefulView name initial buildNode = unsafePerformIO $ do
   stateEq <- singleEq (Proxy :: Proxy state)
   View <$> js_createNewStatefulView name initialRef renderCb propsEq stateEq
 {-# NOINLINE mkStatefulView #-}
-
-foreign import javascript unsafe
-    "hsreact$mk_new_stateful_view($1, $2, $3, $4, $5)"
-    js_createNewStatefulView
-        :: JSString
-        -> Export state
-        -> Callback (JSVal -> JSVal -> IO ())
-        -> Callback (JSVal -> JSVal -> IO JSVal)
-        -> Callback (JSVal -> JSVal -> IO JSVal)
-        -> IO (ReactViewRef props)
 
 mkControllerView :: forall (stores :: [*]) (props :: [*]).
                     (ControllerViewStores stores, Typeable stores, AllEq stores,
@@ -293,16 +267,6 @@ mkControllerView name buildNode = unsafePerformIO $ do
   stateEq <- allEq (Proxy :: Proxy stores)
   View <$> js_createNewCtrlView name renderCb artifacts propsEq stateEq
 {-# NOINLINE mkControllerView #-}
-
-foreign import javascript unsafe
-  "hsreact$mk_new_ctrl_view($1, $2, $3, $4, $5)"
-  js_createNewCtrlView
-    :: JSString
-    -> Callback (JSVal -> JSVal -> IO ())
-    -> Artifacts
-    -> Callback (JSVal -> JSVal -> IO JSVal)
-    -> Callback (JSVal -> JSVal -> IO JSVal)
-    -> IO (ReactViewRef props)
 
 exportReactViewToJavaScript :: forall (props :: [*]). (ExportViewProps props) => View props -> IO JSVal
 exportReactViewToJavaScript (View v) = do
@@ -348,6 +312,64 @@ runStateViewHandler this handler = do
   -- TODO: I'm not so optimistic about the above code not blocking (is that the same as yielding?)
   actions `deepseq` mapM_ executeAction actions
 
+getProp :: Typeable a => NewJsProps -> Int -> IO a
+getProp p i = js_getPropFromList p i >>= unsafeDerefExport "getProp"
+
+pushProp :: Typeable a => a -> NewJsProps -> IO ()
+pushProp val props = do
+  valE <- export $! val -- this will be released in the lifecycle callbacks of the class
+  js_pushProp props valE
+
+findFromState :: Typeable a => Int -> JsState -> IO a
+findFromState i s = js_findFromState i s >>= unsafeDerefExport "findFromState"
+
+newtype Artifacts = Artifacts JSVal
+instance IsJSVal Artifacts
+
+newtype Artifact = Artifact JSVal
+instance IsJSVal Artifact
+
+newtype RenderCbArg = RenderCbArg JSVal
+instance IsJSVal RenderCbArg
+
+#ifdef __GHCJS__
+
+foreign import javascript unsafe
+  "$1.input"
+  js_getDeriveInput :: JSVal -> IO (Export store)
+
+foreign import javascript unsafe
+  "$1.output = $2"
+  js_setDeriveOutput :: JSVal -> Export a -> IO ()
+
+foreign import javascript unsafe
+    "hsreact$mk_new_view($1, $2, $3)"
+    js_createNewView
+        :: JSString
+        -> Callback (JSVal -> JSVal -> IO ())
+        -> Callback (JSVal -> JSVal -> IO JSVal)
+        -> IO (ReactViewRef props)
+
+foreign import javascript unsafe
+    "hsreact$mk_new_stateful_view($1, $2, $3, $4, $5)"
+    js_createNewStatefulView
+        :: JSString
+        -> Export state
+        -> Callback (JSVal -> JSVal -> IO ())
+        -> Callback (JSVal -> JSVal -> IO JSVal)
+        -> Callback (JSVal -> JSVal -> IO JSVal)
+        -> IO (ReactViewRef props)
+
+foreign import javascript unsafe
+  "hsreact$mk_new_ctrl_view($1, $2, $3, $4, $5)"
+  js_createNewCtrlView
+    :: JSString
+    -> Callback (JSVal -> JSVal -> IO ())
+    -> Artifacts
+    -> Callback (JSVal -> JSVal -> IO JSVal)
+    -> Callback (JSVal -> JSVal -> IO JSVal)
+    -> IO (ReactViewRef props)
+
 foreign import javascript unsafe
   "[]"
   js_newEmptyPropList :: IO NewJsProps
@@ -368,26 +390,9 @@ foreign import javascript unsafe
   "$1.push($2)"
   js_pushProp :: NewJsProps -> Export a -> IO ()
 
-getProp :: Typeable a => NewJsProps -> Int -> IO a
-getProp p i = js_getPropFromList p i >>= unsafeDerefExport "getProp"
-
-pushProp :: Typeable a => a -> NewJsProps -> IO ()
-pushProp val props = do
-  valE <- export $! val -- this will be released in the lifecycle callbacks of the class
-  js_pushProp props valE
-
-findFromState :: Typeable a => Int -> JsState -> IO a
-findFromState i s = js_findFromState i s >>= unsafeDerefExport "findFromState"
-
 foreign import javascript unsafe
   "$2[$1]"
   js_findFromState :: Int -> JsState -> IO (Export a)
-
-newtype Artifacts = Artifacts JSVal
-instance IsJSVal Artifacts
-
-newtype Artifact = Artifact JSVal
-instance IsJSVal Artifact
 
 foreign import javascript unsafe
   "{}"
@@ -417,9 +422,83 @@ foreign import javascript unsafe
   "$1._updateAndReleaseState($2)"
   js_ReactUpdateAndReleaseState :: ReactThis state props -> Export state -> IO ()
 
-newtype RenderCbArg = RenderCbArg JSVal
-instance IsJSVal RenderCbArg
-
 foreign import javascript unsafe
   "$1.newCallbacks = $2; $1.elem = $3;"
   js_RenderCbSetResults :: RenderCbArg -> JSVal -> ReactElementRef -> IO ()
+
+#else
+
+js_getDeriveInput :: JSVal -> IO (Export store)
+js_getDeriveInput _ = error "js_getDeriveInput only works with GHCJS"
+
+js_setDeriveOutput :: JSVal -> Export a -> IO ()
+js_setDeriveOutput _ _ = error "js_setDeriveOutput only works with GHCJS"
+
+js_createNewView
+    :: JSString
+    -> Callback (JSVal -> JSVal -> IO ())
+    -> Callback (JSVal -> JSVal -> IO JSVal)
+    -> IO (ReactViewRef props)
+js_createNewView _ _ _ = error "js_createNewView only works with GHCJS"
+
+js_createNewStatefulView
+    :: JSString
+    -> Export state
+    -> Callback (JSVal -> JSVal -> IO ())
+    -> Callback (JSVal -> JSVal -> IO JSVal)
+    -> Callback (JSVal -> JSVal -> IO JSVal)
+    -> IO (ReactViewRef props)
+js_createNewStatefulView _ _ _ _ _ = error "js_createNewStatefulView only works with GHCJS"
+
+js_createNewCtrlView
+    :: JSString
+    -> Callback (JSVal -> JSVal -> IO ())
+    -> Artifacts
+    -> Callback (JSVal -> JSVal -> IO JSVal)
+    -> Callback (JSVal -> JSVal -> IO JSVal)
+    -> IO (ReactViewRef props)
+js_createNewCtrlView _ _ _ _ _ = error "js_createNewCtrlView only works with GHCJS"
+
+js_newEmptyPropList :: IO NewJsProps
+js_newEmptyPropList = error "js_newEmptyPropList only works with GHCJS"
+
+js_PropsList :: ReactThis state props -> IO NewJsProps
+js_PropsList _ = error "js_PropsList only works with GHCJS"
+
+js_NewStateDict :: ReactThis state props -> IO JsState
+js_NewStateDict _ = error "js_NewStateDict only works with GHCJS"
+
+js_getPropFromList :: NewJsProps -> Int -> IO (Export a)
+js_getPropFromList _ _ = error "js_getPropFromList only works with GHCJS"
+
+js_pushProp :: NewJsProps -> Export a -> IO ()
+js_pushProp _ _ = error "js_pushProp only works with GHCJS"
+
+js_findFromState :: Int -> JsState -> IO (Export a)
+js_findFromState _ _ = error "js_findFromState only works with GHCJS"
+
+js_emptyArtifacts :: IO Artifacts
+js_emptyArtifacts = error "js_emptyArtifacts only works with GHCJS"
+
+js_newArtifact :: IO Artifact
+js_newArtifact = error "js_newArtifact only works with GHCJS"
+
+js_addStoreState :: Artifact -> Int -> IO ()
+js_addStoreState _ _ = error "js_addStoreState only works with GHCJS"
+
+js_addStoreDerivedState :: Artifact -> Int -> Callback (JSVal -> IO ()) -> IO ()
+js_addStoreDerivedState _ _ _ = error "js_addStoreDerivedState only works with GHCJS"
+
+js_setArtifact :: Artifacts -> JSString -> Artifact -> IO ()
+js_setArtifact _ _ _ = error "js_setArtifact only works with GHCJS"
+
+js_ReactGetState :: ReactThis state props -> IO (Export state)
+js_ReactGetState _ = error "js_ReactGetState only works with GHCJS"
+
+js_ReactUpdateAndReleaseState :: ReactThis state props -> Export state -> IO ()
+js_ReactUpdateAndReleaseState _ _ = error "js_ReactUpdateAndReleaseState only works with GHCJS"
+
+js_RenderCbSetResults :: RenderCbArg -> JSVal -> ReactElementRef -> IO ()
+js_RenderCbSetResults _ _ _ = error "js_RenderCbSetResults only works with GHCJS"
+
+#endif

--- a/react-hs/test/client/TestClient.hs
+++ b/react-hs/test/client/TestClient.hs
@@ -1,4 +1,4 @@
-{-# LANGUAGE OverloadedStrings, TypeFamilies, ScopedTypeVariables, DeriveAnyClass,
+{-# LANGUAGE CPP, OverloadedStrings, TypeFamilies, ScopedTypeVariables, DeriveAnyClass,
              FlexibleInstances, DeriveGeneric, BangPatterns, TemplateHaskell, DataKinds, TypeApplications, MultiParamTypeClasses #-}
 {-# OPTIONS_GHC -fno-warn-orphans #-}
 module Main (main) where
@@ -18,10 +18,6 @@ import GHCJS.Types (JSVal, JSString)
 import JavaScript.Array (JSArray)
 import qualified Data.JSString.Text as JSS
 
-foreign import javascript unsafe
-    "hsreact$log_message($1)"
-    js_output :: JSString -> IO ()
-
 data OutputStoreData = OutputStoreData
     deriving (Eq, Show, Typeable)
 
@@ -37,10 +33,6 @@ initOutputStore = registerInitialStore OutputStoreData
 
 output :: [T.Text] -> [SomeStoreAction]
 output s = [someStoreAction @OutputStoreData s]
-
-foreign import javascript unsafe
-  "React['createElement']('p', {'id': 'test-raw-js-para', 'key': 'test-raw-para'}, $2)"
-  js_testRawJs :: JSVal -> JSArray -> IO JSVal
 
 --------------------------------------------------------------------------------
 --- Events
@@ -331,10 +323,6 @@ callbackViewWrapper = mkView "callback view wrapper" $
 --- Intl
 --------------------------------------------------------------------------------
 
-foreign import javascript unsafe
-    "{'with_trans': 'message from translation {abc}'}"
-    js_translations :: JSVal
-
 intlSpec :: View '[]
 intlSpec = mkView "intl" $
     intlProvider_ "en" (Just js_translations) Nothing $
@@ -451,6 +439,33 @@ main = do
   initOutputStore
   initCharacterStore
   reactRenderView "app" testClient
+
+#ifdef __GHCJS__
+
+foreign import javascript unsafe
+    "hsreact$log_message($1)"
+    js_output :: JSString -> IO ()
+
+foreign import javascript unsafe
+    "React['createElement']('p', {'id': 'test-raw-js-para', 'key': 'test-raw-para'}, $2)"
+    js_testRawJs :: JSVal -> JSArray -> IO JSVal
+
+foreign import javascript unsafe
+    "{'with_trans': 'message from translation {abc}'}"
+    js_translations :: JSVal
+
+#else
+
+js_output :: JSString -> IO ()
+js_output _ = error "js_output only works with GHCJS"
+
+js_testRawJs :: JSVal -> JSArray -> IO JSVal
+js_testRawJs _ _ = error "js_testRawJs only works with GHCJS"
+
+js_translations :: JSVal
+js_translations = error "js_translations only works with GHCJS"
+
+#endif
 
 writeIntlMessages (intlFormatJson "test/client/msgs/jsonmsgs.json")
 writeIntlMessages (intlFormatJsonWithoutDescription "test/client/msgs/jsonnodescr.json")

